_posts/2014-11-13-cockpit-plugin-tutorial.md

```diff
@@ -78,92 +86,105 @@ Lets take a look at the pinger HTML, and see how it works.
 ```
 
-In the `ping_run()` function is where the magic happens. `cockpit.spawn` is a function, [documented here](https://cockpit-project.org/guide/latest/api-cockpit.html#latest-spawn) that lets you spawn processes on the server and interact with them via stdin and stdout. Here we spawn the `ping` command with some arguments:
+The `ping_run()` function is where the magic happens. [cockpit.spawn](https://cockpit-project.org/guide/latest/cockpit-spawn.html) lets you spawn processes on the server and interact with them via `stdin` and `stdout`.
+
+Here we spawn the `ping` command with some arguments:
 
-```javascript
-    var proc = cockpit.spawn(["ping", "-c", "4", address.val()]);
+```js
+cockpit.spawn(["ping", "-c", "4", address.val()])
 ```
 
-In a web browser you cannot block and wait until a method call completes. Anything that doesn't happen instantaneously gets its results reported back to you by [means of callback handlers](https://cockpit-project.org/guide/latest/api-cockpit.html#latest-spawn-done). jQuery has a standard interface [called a promise](http://api.jquery.com/deferred.promise/). You add handlers by calling the `.done()` or `.fail()` methods and registering callbacks. `proc.stream()` registers a callback to be invoked whenever the process produces output.
+In a web browser you cannot block and wait until a method call completes. Anything that doesn't happen instantaneously gets its results reported back to you by [means of callback handlers](https://cockpit-project.org/guide/latest/cockpit-spawn.html#cockpit-spawn-then).
 
-```javascript
-    proc.done(ping_success);
-    proc.stream(ping_output);
-    proc.fail(ping_fail);
-    ...
+JavaScript has a standard interface called a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise). You add handlers by calling the `.then()` or `.catch()` methods and registering callbacks (Note: historically these have been called `.done()` and `.fail()`, but these should not be used any more in new code.) The `cockpit.spawn` specific `.stream()` handler registers a callback to be invoked whenever the process produces output.
+
+```js
+   <some promise>
+   .stream(ping_output)
+   .then(ping_success)
+   .catch(ping_fail);
 }
 ```
 
 The `ping_success()` and `ping_fail()` and `ping_output()` update the display as you would expect.
 
-So there you go ... it's a simple plugin to start off with ... next time we'll cover [how to use DBus](https://cockpit-project.org/guide/latest/api-cockpit.html#latest-dbus), and then the real fun begins.
+**Warning**: Don't start long-running and uninterruptible processes in this naïve way. Browser tabs get closed, internet connections severed, laptops get suspended, etc. Take care to use a mechanism like [systemd-run](https://www.freedesktop.org/software/systemd/man/systemd-run.html) or similar for use cases like installation processes that should not be interrupted by a closing Cockpit session.
```

Comment:
```suggestion
_**Warning**: Don't start long-running, uninterruptible tasks in this naïve way. Browser tabs are easily closed, Internet connections get severed, laptops suspend, and so on. Take care to use a mechanism like [systemd-run](https://www.freedesktop.org/software/systemd/man/systemd-run.html) (or similar) for processes that should not be interrupted, such as installation procedures._
```

Reply:
We could also use an admonition style here too (in addition to the note at the top of the page).

Don't let either lack of call-out style block this PR though. We should open another issue to handle it and retroactively adjust these (and possibly other blog posts) "note:", "info:", "warning:", etc. notices.